torch/csrc/jit/passes/onnx.cpp

```diff
@@ -355,5 +357,129 @@ void BlockToONNX(
+  struct ToScan {
+    ModulePtr mod;
+    Node* n;
+    size_t offset;
+  };
+  std::vector<ToScan> to_scan;
+  std::vector<Node*> to_clean; // nodes that should be dead at the end
+
+  auto getOrAddSlot = [&](const Slot& slot) -> Value* {
+    auto it = slot_to_offset.find(slot);
+    if (it != slot_to_offset.end()) {
+      size_t ivalues_start = g->inputs().size() - extra_ivalues.size();
+      return g->inputs().at(ivalues_start + it->second);
+    }
+    extra_ivalues.emplace_back(slot);
+    slot_to_offset[slot] = extra_ivalues.size() - 1;
+    return g->addInput()->setType(slot.obj->getSlot(slot.offset).type());
+  };
+
+  auto self_value = g->inputs().at(self_offset);
+
+  for (Use use : self_value->uses()) {
+    to_scan.emplace_back(ToScan{self, use.user, use.offset});
+  }
+  while (to_scan.size() > 0) {
+    auto e = to_scan.back();
+    to_scan.pop_back();
+
+    // when we lambda lift forks, first-class modules may be passed across
+    // forks. This code recursively lowers the module in the fork call.
+    if (e.n->kind() == prim::fork) {
+      auto subgraph = e.n->g(attr::Subgraph);
+      std::vector<Slot> new_slots;
+      std::tie(subgraph, new_slots) = lower_graph(e.mod, *subgraph, e.offset);
+      e.n->g_(attr::Subgraph, subgraph);
+      for (const Slot& slot : new_slots) {
+        e.n->addInput(getOrAddSlot(slot));
+      }
+      e.n->removeInput(e.offset);
+      continue;
+    }
+    if (e.n->kind() == prim::PythonOp) {
+      throw script::ErrorReport(e.n->sourceRange())
+          << "Couldn't export Python method.";
+    }
+    if (e.n->kind() != prim::GetAttr) {
+      throw script::ErrorReport(e.n->sourceRange())
+          << "temporary: the only valid use of a module is looking up an "
+             "attribute but found "
+          << *e.n;
+    }
+    size_t slot_idx = e.mod->type()->getAttributeSlot(e.n->s(attr::name));
+    auto iv = e.mod->getSlot(slot_idx);
+    if (ClassTypePtr c = e.n->output()->type()->cast<ClassType>()) {
+      if (c->is_module()) {
+        for (Use use : e.n->output()->uses()) {
+          to_scan.emplace_back(ToScan{iv.toObject(), use.user, use.offset});
+        }
+        to_clean.emplace_back(e.n);
+        continue;
+      }
+    }
+    e.n->output()->replaceAllUsesWith(getOrAddSlot({e.mod, slot_idx}));
+    e.n->destroy();
+  }
+
+  while (to_clean.size() > 0) {
+    Node* n = to_clean.back();
+    AT_ASSERT(!n->hasUses());
+    n->destroy();
+    to_clean.pop_back();
+  }
+  AT_ASSERT(!self_value->hasUses());
+  g->eraseInput(self_offset);
+
+  return std::make_pair(std::move(g), std::move(extra_ivalues));
+}
+
+static std::vector<at::Tensor> loadTensors(const std::vector<Slot>& slots) {
+  std::vector<at::Tensor> result;
+  result.reserve(slots.size());
+  for (const Slot& slot : slots) {
+    result.emplace_back(slot.obj->getSlot(slot.offset).toTensor());
+  }
+  return result;
+}
+
+std::pair<std::shared_ptr<Graph>, std::vector<at::Tensor>> ONNXLowerGraph(
```

Comment:
Technically this is not ONNX specific so we probably should keep it as LowerGraph. It is probable that some backend integrations are using this function, and will need to be able to find it again after we move it.